### PR TITLE
[optimize](storage)optimize date in storage layer

### LINKED ISSUE #8796: [Proposal] Make date/decimal read and evaluation faster
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Currently date/decimal may cost too many times when evaluating predicate.
Because they are customized class and can not be evaluated in vectorized way.
For decimal, I think we can convert it to int128, then evaluating it in vectorized way. The ```convert```  may be slow in current insert method, but I think we can try to vectorized it.
For date, currently implemented using a 24-byte array, I think maybe we can convert it to a integer and then evaluate it in vectorized way.

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wangbo**
# Proposed changes

Issue Number: close #8796

## Problem Summary:

1 using uint32_t to evaluate pred for date type.
2 remove range check when storage output date type.

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by wangbo** [comment]
#  Performance Test
## SQL1
```
SELECT sum(LO_EXTENDEDPRICE * LO_DISCOUNT) AS revenue
FROM lineorder_flat
WHERE LO_ORDERDATE >= 19930101 and LO_ORDERDATE <= 19931231 AND LO_DISCOUNT BETWEEN 1 AND 3 AND LO_QUANTITY < 25;

before:
- BlockLoadTime: 1s226ms
- PredColumnReadTime: 734.907ms
- VectorPredEvalTime: 296.752ms
- OutputColumnTime: 179.064ms

after:
- BlockLoadTime: 903.229ms
- PredColumnReadTime: 738.852ms
- VectorPredEvalTime: 83.223ms
- OutputColumnTime: 65.080ms
```

## SQL2:  pred only contains date
```
SELECT sum(LO_DISCOUNT) AS revenue
FROM lineorder_flat
WHERE LO_ORDERDATE >= 19930101 and LO_ORDERDATE <= 19931231

before:
- BlockLoadTime: 1s982ms
- PredColumnReadTime: 478.911ms
- VectorPredEvalTime: 307.892ms
- OutputColumnTime: 1s180ms

after:
- BlockLoadTime: 879.068ms
- PredColumnReadTime: 489.480ms
- VectorPredEvalTime: 100.503ms
- OutputColumnTime: 274.997ms

```

## SQL3: no predicate
```
select count(distinct LO_ORDERDATE) from lineorder_flat

before:
- BlockLoadTime: 8s395ms
- PredColumnReadTime: 8s002ms

after:
- BlockLoadTime: 6s641ms
- PredColumnReadTime: 6s252ms
```

We can see ```VectorPredEvalTime```  has significant performance improvement.

File: be/src/vec/runtime/vdatetime_value.h [hunk 1]
```diff
@@ -203,6 +203,21 @@ class VecDateTimeValue { // Now this type is a temp solution with little changes
...
+    //note(wb) not check in this method
+    void inline set_olap_date(uint64_t olap_date_val) {
+        _neg = 0;
+        _type = TIME_DATE;
+
+         _day = olap_date_val & 0x1f;
+        _month = (olap_date_val >> 5) & 0x0f;
+        _year = olap_date_val >> 9;
+        _neg = 0;
+        _type = TIME_DATE;
```
Comment (zenoyang) [inline be/src/vec/runtime/vdatetime_value.h:215]:
```
_neg = 0;
_type = TIME_DATE;
```
Duplicated with the code above

File: be/src/vec/columns/predicate_column.h [hunk 2]
```diff
@@ -205,13 +218,30 @@ class PredicateColumnType final : public COWHelper<IColumn, PredicateColumnType<
...
     }
 
+    void insert_many_date(const char* data_ptr, size_t num) {
+        size_t intput_type_size  = sizeof(uint24_t);
+        size_t res_type_size = sizeof(uint32_t);
+        char* input_data_ptr = const_cast<char*>(data_ptr);
+        
+        char* res_ptr = (char*)data.get_end_ptr();
+        memset(res_ptr, 0, res_type_size * num);
+        for (int i = 0; i < num; i++) {
```
Comment (zenoyang) [inline be/src/vec/columns/predicate_column.h:228]: Why not batch memcpy?

File: be/src/vec/columns/predicate_column.h [hunk 3]
```diff
@@ -64,6 +64,19 @@ class PredicateColumnType final : public COWHelper<IColumn, PredicateColumnType<
         }
     }
 
+    void insert_date32_to_res_column(const uint16_t* sel, size_t sel_size,
+                                   vectorized::ColumnVector<Int64>* res_ptr) {
+        res_ptr->reserve(sel_size);
+        auto& res_data = res_ptr->get_data();
+
+        for (size_t i = 0; i < sel_size; i++) {
+            VecDateTimeValue date;
```
Comment (zenoyang) [inline be/src/vec/columns/predicate_column.h:73]: same as above

File: be/src/vec/columns/column_vector.h [hunk 4]
```diff
@@ -168,18 +168,16 @@ class ColumnVector final : public COWHelper<ColumnVectorHelper, ColumnVector<T>>
...
-            value <<= 8;
-            value |= *(unsigned char*)(cur_ptr);
-            vectorized::VecDateTimeValue date;
-            date.from_olap_date(value);
-            this->insert_data(reinterpret_cast<char*>(&date), 0);
+            uint64_t val = 0;
+            memcpy((char*)(&val), data_ptr, input_value_size);
+            data_ptr += input_value_size;
+
+            VecDateTimeValue date;
```
Comment (zenoyang) [inline be/src/vec/columns/column_vector.h:178]: ```VecDateTimeValue date;```
Better to move outside the loop.

File: be/src/vec/columns/predicate_column.h (hunk 2 above)
Comment (wangbo) [reply be/src/vec/columns/predicate_column.h:228]: ```input_data_ptr``` is a uint24_t array.
```res_ptr``` is a uint32_t array.
The element length is different.

File: be/src/vec/columns/column_vector.h (hunk 4 above)
Comment (wangbo) [reply be/src/vec/columns/column_vector.h:178]:
```
q1
- OutputColumnTime: 90.720ms
q2
- OutputColumnTime: 286.477ms
q3
- PredColumnReadTime: 6s827ms
```
I did test after move it out of loop.
There is no noticeable performance change.
I guess  the cost of allocate a Object like ```VecDateTimeValue```  is similar to allocate a int64.

File: be/src/vec/columns/predicate_column.h (hunk 2 above)
Comment (zenoyang) [reply be/src/vec/columns/predicate_column.h:228]: ok

File: be/src/vec/columns/column_vector.h (hunk 4 above)
Comment (zenoyang) [reply be/src/vec/columns/column_vector.h:178]: ok

File: be/src/vec/columns/predicate_column.h [hunk 5]
```diff
@@ -405,6 +435,9 @@ class PredicateColumnType final : public COWHelper<IColumn, PredicateColumnType<
         } else if constexpr (std::is_same_v<T, uint24_t>) {
```
Comment (Gabriel39) [inline be/src/vec/columns/predicate_column.h:435]: could this code path be deleted?

Comment (wangbo) [reply be/src/vec/columns/predicate_column.h:435]: No, compile could failed.

File: be/src/vec/runtime/vdatetime_value.h [hunk 6]
```diff
@@ -223,6 +223,19 @@ class VecDateTimeValue { // Now this type is a temp solution with little changes
         return check_range_and_set_time(year, month, day, hour, minute, second, _type);
     }
 
+    //note(wb) not check in this method
```
[the remark this answers is not in the document]
Comment (wangbo) [reply be/src/vec/runtime/vdatetime_value.h:226]: Only used for read from storage layer is ok.

File: be/src/vec/columns/column_vector.h [hunk 7]
```diff
@@ -169,17 +169,16 @@ class ColumnVector final : public COWHelper<ColumnVectorHelper, ColumnVector<T>>
     }
 
     void insert_date_column(const char* data_ptr, size_t num) {
-        size_t value_size = sizeof(uint24_t);
+        size_t input_value_size = sizeof(uint24_t);
+        
         for (int i = 0; i < num; i++) {
```
[the remark this answers is not in the document]
Comment (wangbo) [reply be/src/vec/columns/column_vector.h:174]: What's the difference?

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**